File: ribo_switch/fold.py
import math
import numpy as np
from dataclasses import dataclass
from ribo_switch.types import Base, Energy, Sequence, CANONICAL_PAIRS
from ribo_switch.turner import TurnerParams, INF

@dataclass
class FoldResult:
    mfe_energy: Energy
    mfe_structure: str
    v: np.ndarray
    w: np.ndarray
MIN_HAIRPIN = 3

def fold_mfe(seq: Sequence, params: TurnerParams) -> FoldResult:
    n = len(seq)
    v = np.full((n, n), INF, dtype=np.int64)
    wm = np.full((n, n), INF, dtype=np.int64)
    for span in range(MIN_HAIRPIN + 2, n + 1):
        for i in range(n - span + 1):
            j = i + span - 1
            if _can_pair(seq, i, j):
                v[i][j] = _fill_v(seq, params, v, wm, i, j)
            wm[i][j] = _fill_wm(seq, params, v, wm, i, j)
    f5 = _fill_f5(seq, params, v, n)
    mfe = int(f5[n])
    structure = _traceback(seq, params, v, f5, wm, n)
    return FoldResult(mfe_energy=mfe, mfe_structure=structure, v=v, w=f5)

def _fill_f5(seq: Sequence, params: TurnerParams, v: np.ndarray, n: int) -> np.ndarray:
    f5 = np.zeros(n + 1, dtype=np.int64)
    for i in range(1, n + 1):
        best = int(f5[i - 1])
        for k in range(0, i):
            if i - 1 - k < MIN_HAIRPIN + 1:
                continue
            if not _can_pair(seq, k, i - 1):
                continue
            if v[k][i - 1] >= INF:
                continue
            pi = _pair_index(seq, k, i - 1)
            e = int(f5[k]) + int(v[k][i - 1])
            if pi in (0, 1, 4, 5):
                e += params.terminal_au_penalty
            if k > 0:
                e += int(params.dangle5[pi][seq.bases[k - 1].value])
            if i < n:
                e += int(params.dangle3[pi][seq.bases[i].value])
            if e < best:
                best = e
        f5[i] = best
    return f5

def _can_pair(seq: Sequence, i: int, j: int) -> bool:
    return (seq.bases[i], seq.bases[j]) in CANONICAL_PAIRS

def _pair_index(seq: Sequence, i: int, j: int) -> int:
    bi, bj = (seq.bases[i], seq.bases[j])
    _MAP = {(Base.A, Base.U): 0, (Base.U, Base.A): 1, (Base.C, Base.G): 2, (Base.G, Base.C): 3, (Base.G, Base.U): 4, (Base.U, Base.G): 5}
    return _MAP[bi, bj]

def _fill_v(seq: Sequence, params: TurnerParams, v: np.ndarray, wm: np.ndarray, i: int, j: int) -> int:
    best = INF
    n = len(seq)
    pair_ij = _pair_index(seq, i, j)
    best = min(best, _hairpin_energy(seq, params, i, j, pair_ij))
    max_interior = min(j - i - 2, 30)
    for p in range(i + 1, j):
        n_left = p - i - 1
        if n_left > max_interior:
            break
        for q in range(j - 1, p, -1):
            n_right = j - q - 1
            total_unp = n_left + n_right
            if total_unp > max_interior:
                continue
            if not _can_pair(seq, p, q):
                continue
            if q - p - 1 < MIN_HAIRPIN:
                continue
            pair_pq = _pair_index(seq, p, q)
            if n_left == 0 and n_right == 0:
                e = int(params.stack[pair_ij][pair_pq]) + int(v[p][q])
                best = min(best, e)
            elif n_left == 0 or n_right == 0:
                bulge_size = n_left + n_right
                e = _bulge_e(params, pair_ij, pair_pq, bulge_size) + int(v[p][q])
                best = min(best, e)
            else:
                e = _interior_e(seq, params, i, j, p, q, pair_ij, pair_pq, n_left, n_right) + int(v[p][q])
                best = min(best, e)
    if j - i - 1 >= 2 * (MIN_HAIRPIN + 2):
        ml_e = params.ml_offset + params.ml_per_branch
        if pair_ij in (0, 1, 4, 5):
            ml_e += params.terminal_au_penalty
        for k in range(i + 2 + MIN_HAIRPIN, j - MIN_HAIRPIN - 1):
            branch_e = ml_e + int(wm[i + 1][k]) + int(wm[k + 1][j - 1])
            best = min(best, branch_e)
    return best

def _fill_wm(seq: Sequence, params: TurnerParams, v: np.ndarray, wm: np.ndarray, i: int, j: int) -> int:
    best = INF
    if _can_pair(seq, i, j) and v[i][j] < INF:
        branch_e = int(v[i][j]) + params.ml_per_branch
        pair_idx = _pair_index(seq, i, j)
        if pair_idx in (0, 1, 4, 5):
            branch_e += params.terminal_au_penalty
        best = min(best, branch_e)
    if i + 1 <= j and wm[i + 1][j] < INF:
        best = min(best, int(wm[i + 1][j]) + params.ml_per_unpaired)
    if i <= j - 1 and wm[i][j - 1] < INF:
        best = min(best, int(wm[i][j - 1]) + params.ml_per_unpaired)
    for k in range(i + 1, j):
        if wm[i][k] < INF and wm[k + 1][j] < INF:
            best = min(best, int(wm[i][k]) + int(wm[k + 1][j]))
    return best

def _hairpin_energy(seq: Sequence, params: TurnerParams, i: int, j: int, pair_idx: int) -> int:
    size = j - i - 1
    if size < MIN_HAIRPIN:
        return INF
    if size <= 30:
        energy = int(params.hairpin_init[size])
    else:
        energy = int(params.hairpin_init[30]) + int(round(params.loop_extrapolation_coeff * 100 * math.log(size / 30.0)))
    if size == 3:
        loop_seq = ''.join((seq.bases[k].name for k in range(i, j + 1)))
        bonus = params.hairpin_triloop.get(loop_seq, 0)
        if bonus:
            energy = bonus
        if pair_idx in (0, 1, 4, 5):
            energy += params.terminal_au_penalty
        if all((seq.bases[k] == Base.C for k in range(i + 1, j))):
            energy += params.hairpin_c3
        return energy
    if size == 4:
        loop_seq = ''.join((seq.bases[k].name for k in range(i, j + 1)))
        bonus = params.hairpin_tetraloop.get(loop_seq, 0)
        if bonus:
            return bonus
    b5 = seq.bases[i + 1].value
    b3 = seq.bases[j - 1].value
    energy += int(params.hairpin_mismatch[pair_idx][b5][b3])
    mm = (seq.bases[i + 1], seq.bases[j - 1])
    if mm in ((Base.U, Base.U), (Base.G, Base.A)):
        energy += params.hairpin_uu_ga_bonus
    if mm == (Base.G, Base.G):
        energy += params.hairpin_gg_bonus
    if pair_idx == 4:
        if i >= 2 and seq.bases[i - 1] == Base.G and (seq.bases[i - 2] == Base.G):
            energy += params.hairpin_special_gu
    if all((seq.bases[k] == Base.C for k in range(i + 1, j))):
        energy += params.hairpin_c_slope * size + params.hairpin_c_intercept
    return energy

def _bulge_e(params: TurnerParams, pair_ij: int, pair_pq: int, size: int) -> int:
    if size <= 30:
        energy = int(params.bulge_init[size])
    else:
        energy = int(params.bulge_init[30]) + int(round(params.loop_extrapolation_coeff * 100 * math.log(size / 30.0)))
    if size == 1:
        energy += int(params.stack[pair_ij][pair_pq])
    if pair_ij in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    if pair_pq in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    return energy

def _interior_e(seq: Sequence, params: TurnerParams, i: int, j: int, p: int, q: int, pair_ij: int, pair_pq: int, n_left: int, n_right: int) -> int:
    total = n_left + n_right
    if n_left == 1 and n_right == 1:
        mm5 = seq.bases[i + 1].value
        mm3 = seq.bases[j - 1].value
        val = int(params.interior_1x1[pair_ij][pair_pq][mm5][mm3])
        if val < INF:
            return val
    if total <= 30:
        energy = int(params.interior_init[total])
    else:
        energy = int(params.interior_init[30]) + int(round(params.loop_extrapolation_coeff * 100 * math.log(total / 30.0)))
    asym = abs(n_left - n_right)
    energy += min(params.ninio_max, params.ninio_m * asym)
    if not (n_left == 1 and n_right == 1):
        b5o = seq.bases[i + 1].value
        b3o = seq.bases[j - 1].value
        energy += int(params.interior_mismatch[pair_ij][b5o][b3o])
        b3i = seq.bases[p - 1].value if p > 0 else 0
        b5i = seq.bases[q + 1].value if q + 1 < len(seq) else 0
        energy += int(params.interior_mismatch[pair_pq][b3i][b5i])
    if pair_ij in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    if pair_pq in (0, 1, 4, 5):
        energy += params.terminal_au_penalty
    return energy

def _traceback(seq: Sequence, params: TurnerParams, v: np.ndarray, f5: np.ndarray, wm: np.ndarray, n: int) -> str:
    pairs: list[tuple[int, int]] = []
    _trace_f5(seq, params, v, f5, wm, n, pairs)
    db = ['.'] * n
    for i, j in pairs:
        db[i] = '('
        db[j] = ')'
    return ''.join(db)

def _trace_f5(seq: Sequence, params: TurnerParams, v: np.ndarray, f5: np.ndarray, wm: np.ndarray, n: int, pairs: list[tuple[int, int]]) -> None:
    i = n
    while i > 0:
        if int(f5[i]) == int(f5[i - 1]):
            i -= 1
            continue
        matched = False
        for k in range(0, i):
            if i - 1 - k < MIN_HAIRPIN + 1:
                continue
            if not _can_pair(seq, k, i - 1):
                continue
            if v[k][i - 1] >= INF:
                continue
            pi = _pair_index(seq, k, i - 1)
            e = int(f5[k]) + int(v[k][i - 1])
            if pi in (0, 1, 4, 5):
                e += params.terminal_au_penalty
            if k > 0:
                e += int(params.dangle5[pi][seq.bases[k - 1].value])
            if i < n:
                e += int(params.dangle3[pi][seq.bases[i].value])
            if e == int(f5[i]):
                pairs.append((k, i - 1))
                _trace_v(seq, params, v, f5, wm, k, i - 1, pairs)
                i = k
                matched = True
                break
        if not matched:
            i -= 1

def _trace_v(seq: Sequence, params: TurnerParams, v: np.ndarray, w: np.ndarray, wm: np.ndarray, i: int, j: int, pairs: list[tuple[int, int]]) -> None:
    target = int(v[i][j])
    pair_ij = _pair_index(seq, i, j)
    hp_e = _hairpin_energy(seq, params, i, j, pair_ij)
    if hp_e == target:
        return
    max_interior = min(j - i - 2, 30)
    for p in range(i + 1, j):
        n_left = p - i - 1
        if n_left > max_interior:
            break
        for q in range(j - 1, p, -1):
            n_right = j - q - 1
            total_unp = n_left + n_right
            if total_unp > max_interior:
                continue
            if not _can_pair(seq, p, q):
                continue
            if q - p - 1 < MIN_HAIRPIN:
                continue
            pair_pq = _pair_index(seq, p, q)
            if n_left == 0 and n_right == 0:
                e = int(params.stack[pair_ij][pair_pq]) + int(v[p][q])
            elif n_left == 0 or n_right == 0:
                bulge_size = n_left + n_right
                e = _bulge_e(params, pair_ij, pair_pq, bulge_size) + int(v[p][q])
            else:
                e = _interior_e(seq, params, i, j, p, q, pair_ij, pair_pq, n_left, n_right) + int(v[p][q])
            if e == target:
                pairs.append((p, q))
                _trace_v(seq, params, v, w, wm, p, q, pairs)
                return
    if j - i - 1 >= 2 * (MIN_HAIRPIN + 2):
        ml_e = params.ml_offset + params.ml_per_branch
        if pair_ij in (0, 1, 4, 5):
            ml_e += params.terminal_au_penalty
        for k in range(i + 2 + MIN_HAIRPIN, j - MIN_HAIRPIN - 1):
            if ml_e + int(wm[i + 1][k]) + int(wm[k + 1][j - 1]) == target:
                _trace_wm(seq, params, v, w, wm, i + 1, k, pairs)
                _trace_wm(seq, params, v, w, wm, k + 1, j - 1, pairs)
                return

def _trace_wm(seq: Sequence, params: TurnerParams, v: np.ndarray, w: np.ndarray, wm: np.ndarray, i: int, j: int, pairs: list[tuple[int, int]]) -> None:
    if i > j:
        return
    target = int(wm[i][j])
    if target >= INF:
        return
    if _can_pair(seq, i, j) and v[i][j] < INF:
        branch_e = int(v[i][j]) + params.ml_per_branch
        pair_idx = _pair_index(seq, i, j)
        if pair_idx in (0, 1, 4, 5):
            branch_e += params.terminal_au_penalty
        if branch_e == target:
            pairs.append((i, j))
            _trace_v(seq, params, v, w, wm, i, j, pairs)
            return
    if i + 1 <= j and wm[i + 1][j] < INF:
        if int(wm[i + 1][j]) + params.ml_per_unpaired == target:
            _trace_wm(seq, params, v, w, wm, i + 1, j, pairs)
            return
    if i <= j - 1 and wm[i][j - 1] < INF:
        if int(wm[i][j - 1]) + params.ml_per_unpaired == target:
            _trace_wm(seq, params, v, w, wm, i, j - 1, pairs)
            return
    for k in range(i + 1, j):
        if wm[i][k] < INF and wm[k + 1][j] < INF:
            if int(wm[i][k]) + int(wm[k + 1][j]) == target:
                _trace_wm(seq, params, v, w, wm, i, k, pairs)
                _trace_wm(seq, params, v, w, wm, k + 1, j, pairs)
                return

File: ribo_rs/src/fold.rs
use crate::params::*;
const INF64: i64 = 10_000_000_000;
pub fn fold_mfe(seq: &[u8]) -> (i32, String) {
    let (mfe, _pairs, db) = fold_mfe_full(seq);
    (mfe, db)
}
pub fn fold_mfe_full(seq: &[u8]) -> (i32, Vec<i32>, String) {
    let n = seq.len();
    if n < MIN_HAIRPIN + 2 {
        return (0, vec![-1; n], ".".repeat(n));
    }
    let mut v  = vec![INF64; n * n];
    let mut wm = vec![INF64; n * n];
    macro_rules! V  { ($i:expr,$j:expr) => { v [$i*n+$j] } }
    macro_rules! WM { ($i:expr,$j:expr) => { wm[$i*n+$j] } }
    for span in (MIN_HAIRPIN + 2)..=n {
        for i in 0..=(n - span) {
            let j = i + span - 1;
            if can_pair(seq, i, j) {
                V!(i, j) = fill_v(seq, &v, &wm, i, j, n);
            }
            WM!(i, j) = fill_wm(seq, &v, &wm, i, j, n);
        }
    }
    let f5 = fill_f5(seq, &v, n);
    let mfe = f5[n] as i32;
    let mut pairs = vec![-1i32; n];
    trace_f5(seq, &v, &wm, &f5, n, &mut pairs);
    let mut db = vec![b'.'; n];
    for i in 0..n {
        if pairs[i] > i as i32 {
            db[i] = b'(';
            db[pairs[i] as usize] = b')';
        }
    }
    (mfe, pairs, String::from_utf8(db).unwrap())
}
fn fill_f5(seq: &[u8], v: &[i64], n: usize) -> Vec<i64> {
    let mut f5 = vec![0i64; n + 1];
    for i in 1..=n {
        let mut best = f5[i - 1];
        for k in 0..i {
            if i - 1 - k < MIN_HAIRPIN + 1 { continue; }
            let v_ki = v[k * n + (i - 1)];
            if v_ki >= INF64 { continue; }
            let pi = match pair_index(seq[k], seq[i - 1]) {
                Some(x) => x, None => continue,
            };
            let mut e = f5[k] + v_ki;
            if is_au_gu(pi) { e += TERMINAL_AU_PENALTY as i64; }
            if k > 0 { e += DANGLE5[pi][seq[k - 1] as usize] as i64; }
            if i < n { e += DANGLE3[pi][seq[i] as usize] as i64; }
            if e < best { best = e; }
        }
        f5[i] = best;
    }
    f5
}
fn fill_v(seq: &[u8], v: &[i64], wm: &[i64], i: usize, j: usize, n: usize) -> i64 {
    let pi = match pair_index(seq[i], seq[j]) { Some(x) => x, None => return INF64 };
    let mut best = hairpin64(seq, i, j, pi);
    let max_int = (j - i - 2).min(30);
    for p in i + 1..j {
        let nl = p - i - 1;
        if nl > max_int { break; }
        for q in (p + 1..j).rev() {
            let nr = j - q - 1;
            if nl + nr > max_int { continue; }
            if !can_pair(seq, p, q) { continue; }
            if q - p - 1 < MIN_HAIRPIN { continue; }
            let vp = v[p * n + q];
            if vp >= INF64 { continue; }
            let pp = match pair_index(seq[p], seq[q]) { Some(x) => x, None => continue };
            let e = if nl == 0 && nr == 0 {
                STACK[pi][pp] as i64 + vp
            } else if nl == 0 || nr == 0 {
                bulge64(pi, pp, nl + nr) + vp
            } else {
                interior64(seq, i, j, p, q, pi, pp, nl, nr) + vp
            };
            if e < best { best = e; }
        }
    }
    if j - i - 1 >= 2 * (MIN_HAIRPIN + 2) {
        let mut ml_base = (ML_OFFSET + ML_PER_BRANCH) as i64;
        if is_au_gu(pi) { ml_base += TERMINAL_AU_PENALTY as i64; }
        for k in (i + 2 + MIN_HAIRPIN)..(j - MIN_HAIRPIN - 1) {
            let wml = wm[(i + 1) * n + k];
            let wmr = wm[(k + 1) * n + j - 1];
            if wml < INF64 && wmr < INF64 {
                let e = ml_base + wml + wmr;
                if e < best { best = e; }
            }
        }
    }
    best
}
fn fill_wm(seq: &[u8], v: &[i64], wm: &[i64], i: usize, j: usize, n: usize) -> i64 {
    let mut best = INF64;
    if can_pair(seq, i, j) {
        let vi = v[i * n + j];
        if vi < INF64 {
            let pi = pair_index(seq[i], seq[j]).unwrap();
            let mut e = vi + ML_PER_BRANCH as i64;
            if is_au_gu(pi) { e += TERMINAL_AU_PENALTY as i64; }
            if e < best { best = e; }
        }
    }
    if i + 1 <= j {
        let e = wm[(i + 1) * n + j];
        if e < INF64 {
            let e2 = e + ML_PER_UNPAIRED as i64;
            if e2 < best { best = e2; }
        }
    }
    if j >= 1 {
        let e = wm[i * n + j - 1];
        if e < INF64 {
            let e2 = e + ML_PER_UNPAIRED as i64;
            if e2 < best { best = e2; }
        }
    }
    for k in i + 1..j {
        let el = wm[i * n + k];
        let er = wm[(k + 1) * n + j];
        if el < INF64 && er < INF64 {
            let e = el + er;
            if e < best { best = e; }
        }
    }
    best
}
fn trace_f5(
    seq: &[u8], v: &[i64], wm: &[i64], f5: &[i64],
    n: usize, pairs: &mut Vec<i32>,
) {
    let mut i = n;
    while i > 0 {
        if f5[i] == f5[i - 1] {
            i -= 1;
            continue;
        }
        let mut matched = false;
        for k in 0..i {
            if i - 1 - k < MIN_HAIRPIN + 1 { continue; }
            let v_ki = v[k * n + (i - 1)];
            if v_ki >= INF64 { continue; }
            let pi = match pair_index(seq[k], seq[i - 1]) {
                Some(x) => x, None => continue,
            };
            let mut e = f5[k] + v_ki;
            if is_au_gu(pi) { e += TERMINAL_AU_PENALTY as i64; }
            if k > 0 { e += DANGLE5[pi][seq[k - 1] as usize] as i64; }
            if i < n { e += DANGLE3[pi][seq[i] as usize] as i64; }
            if e == f5[i] {
                pairs[k] = (i - 1) as i32;
                pairs[i - 1] = k as i32;
                trace_v(seq, v, wm, k, i - 1, n, pairs);
                i = k;
                matched = true;
                break;
            }
        }
        if !matched {
            i -= 1;
        }
    }
}
fn trace_v(
    seq: &[u8], v: &[i64], wm: &[i64],
    i: usize, j: usize, n: usize,
    pairs: &mut Vec<i32>,
) {
    let target = v[i * n + j];
    let pi = match pair_index(seq[i], seq[j]) { Some(x) => x, None => return };
    if hairpin64(seq, i, j, pi) == target { return; }
    let max_int = (j - i - 2).min(30);
    for p in i + 1..j {
        let nl = p - i - 1;
        if nl > max_int { break; }
        for q in (p + 1..j).rev() {
            let nr = j - q - 1;
            if nl + nr > max_int { continue; }
            if !can_pair(seq, p, q) { continue; }
            if q - p - 1 < MIN_HAIRPIN { continue; }
            let vp = v[p * n + q];
            if vp >= INF64 { continue; }
            let pp = match pair_index(seq[p], seq[q]) { Some(x) => x, None => continue };
            let e = if nl == 0 && nr == 0 {
                STACK[pi][pp] as i64 + vp
            } else if nl == 0 || nr == 0 {
                bulge64(pi, pp, nl + nr) + vp
            } else {
                interior64(seq, i, j, p, q, pi, pp, nl, nr) + vp
            };
            if e == target {
                pairs[p] = q as i32;
                pairs[q] = p as i32;
                trace_v(seq, v, wm, p, q, n, pairs);
                return;
            }
        }
    }
    if j - i - 1 >= 2 * (MIN_HAIRPIN + 2) {
        let mut ml_base = (ML_OFFSET + ML_PER_BRANCH) as i64;
        if is_au_gu(pi) { ml_base += TERMINAL_AU_PENALTY as i64; }
        for k in (i + 2 + MIN_HAIRPIN)..(j - MIN_HAIRPIN - 1) {
            let wml = wm[(i + 1) * n + k];
            let wmr = wm[(k + 1) * n + j - 1];
            if wml < INF64 && wmr < INF64 && ml_base + wml + wmr == target {
                trace_wm(seq, v, wm, i + 1, k, n, pairs);
                trace_wm(seq, v, wm, k + 1, j - 1, n, pairs);
                return;
            }
        }
    }
}
fn trace_wm(
    seq: &[u8], v: &[i64], wm: &[i64],
    i: usize, j: usize, n: usize,
    pairs: &mut Vec<i32>,
) {
    if i > j { return; }
    let target = wm[i * n + j];
    if target >= INF64 { return; }
    if can_pair(seq, i, j) {
        let vi = v[i * n + j];
        if vi < INF64 {
            let pi = pair_index(seq[i], seq[j]).unwrap();
            let mut e = vi + ML_PER_BRANCH as i64;
            if is_au_gu(pi) { e += TERMINAL_AU_PENALTY as i64; }
            if e == target {
                pairs[i] = j as i32;
                pairs[j] = i as i32;
                trace_v(seq, v, wm, i, j, n, pairs);
                return;
            }
        }
    }
    if i + 1 <= j {
        let e = wm[(i + 1) * n + j];
        if e < INF64 && e + ML_PER_UNPAIRED as i64 == target {
            trace_wm(seq, v, wm, i + 1, j, n, pairs);
            return;
        }
    }
    if j >= 1 {
        let e = wm[i * n + j - 1];
        if e < INF64 && e + ML_PER_UNPAIRED as i64 == target {
            trace_wm(seq, v, wm, i, j - 1, n, pairs);
            return;
        }
    }
    for k in i + 1..j {
        let el = wm[i * n + k];
        let er = wm[(k + 1) * n + j];
        if el < INF64 && er < INF64 && el + er == target {
            trace_wm(seq, v, wm, i, k, n, pairs);
            trace_wm(seq, v, wm, k + 1, j, n, pairs);
            return;
        }
    }
}
#[inline]
fn can_pair(seq: &[u8], i: usize, j: usize) -> bool {
    pair_index(seq[i], seq[j]).is_some()
}
#[inline]
fn hairpin64(seq: &[u8], i: usize, j: usize, pi: usize) -> i64 {
    let e = hairpin_e(seq, i, j, pi);
    if e >= INF { INF64 } else { e as i64 }
}
fn hairpin_e(seq: &[u8], i: usize, j: usize, pi: usize) -> i32 {
    let size = j - i - 1;
    if size < MIN_HAIRPIN { return INF; }
    let mut energy = if size <= 30 {
        HAIRPIN_INIT[size]
    } else {
        HAIRPIN_INIT[30]
            + (LOOP_EXTRAPOLATION_COEFF * (size as f64 / 30.0).ln() * 100.0).round() as i32
    };
    if size == 3 {
        let w = &seq[i..=j];
        let bonus = triloop_bonus(w);
        if bonus != 0 { energy = bonus; }
        if is_au_gu(pi) { energy += TERMINAL_AU_PENALTY; }
        if (i + 1..j).all(|k| seq[k] == 1) { energy += HAIRPIN_C3; }
        return energy;
    }
    if size == 4 {
        let w = &seq[i..=j];
        let bonus = tetraloop_bonus(w);
        if bonus != 0 { return bonus; }
    }
    let b5 = seq[i + 1] as usize;
    let b3 = seq[j - 1] as usize;
    energy += HAIRPIN_MM[pi][b5][b3];
    let mm = (seq[i + 1], seq[j - 1]);
    if mm == (3, 3) || mm == (2, 0) { energy += HAIRPIN_UU_GA_BONUS; }
    if mm == (2, 2) { energy += HAIRPIN_GG_BONUS; }
    if pi == 4 && i >= 2 && seq[i - 1] == 2 && seq[i - 2] == 2 { energy += HAIRPIN_SPECIAL_GU; }
    if (i + 1..j).all(|k| seq[k] == 1) {
        energy += HAIRPIN_C_SLOPE * size as i32 + HAIRPIN_C_INTERCEPT;
    }
    energy
}
#[inline]
fn bulge64(pi: usize, pp: usize, size: usize) -> i64 {
    let mut e = if size <= 30 { BULGE_INIT[size] } else {
        BULGE_INIT[30] + (LOOP_EXTRAPOLATION_COEFF * (size as f64 / 30.0).ln() * 100.0).round() as i32
    };
    if size == 1 { e += STACK[pi][pp]; }
    if is_au_gu(pi) { e += TERMINAL_AU_PENALTY; }
    if is_au_gu(pp) { e += TERMINAL_AU_PENALTY; }
    e as i64
}
#[inline]
fn interior64(
    seq: &[u8],
    i: usize, j: usize,
    p: usize, q: usize,
    pi: usize, pp: usize,
    nl: usize, nr: usize,
) -> i64 {
    let total = nl + nr;
    if nl == 1 && nr == 1 {
        let mm5 = seq[i + 1] as usize;
        let mm3 = seq[j - 1] as usize;
        let val = INT11[pi][pp][mm5][mm3];
        if val < INF { return val as i64; }
    }
    let mut e = if total <= 30 { INTERIOR_INIT[total] } else {
        INTERIOR_INIT[30] + (LOOP_EXTRAPOLATION_COEFF * (total as f64 / 30.0).ln() * 100.0).round() as i32
    };
    let asym = (nl as i32 - nr as i32).unsigned_abs() as i32;
    e += (NINIO_M * asym).min(NINIO_MAX);
    if !(nl == 1 && nr == 1) {
        let b5o = seq[i + 1] as usize;
        let b3o = seq[j - 1] as usize;
        e += INTERIOR_MM[pi][b5o][b3o];
        let b3i = if p > 0 { seq[p - 1] as usize } else { 0 };
        let b5i = if q + 1 < seq.len() { seq[q + 1] as usize } else { 0 };
        e += INTERIOR_MM[pp][b3i][b5i];
    }
    if is_au_gu(pi) { e += TERMINAL_AU_PENALTY; }
    if is_au_gu(pp) { e += TERMINAL_AU_PENALTY; }
    e as i64
}
